bin/differential_tester_x86/LiftAndCompare.cpp

```diff
@@ -0,0 +1,541 @@
+    };
+    test_runner::ExecuteLiftedFunction<X86State, uint64_t>(
+        f1, insn_length, func1_state, mem_handler.get(), pc_fetch);
+    auto second_handler = std::make_unique<test_runner::MemoryHandler>(
+        this->endian, mem_handler->GetUninitializedReads());
+    test_runner::ExecuteLiftedFunction<X86State, uint64_t>(
+        f2, insn_length, func2_state, second_handler.get(), pc_fetch);
+
+
+    auto memory_state_eq =
+        mem_handler->GetMemory() == second_handler->GetMemory();
+
+    if (!memory_state_eq) {
+      LOG(ERROR) << "Memory state differs";
+      LOG(ERROR) << mem_handler->DumpState();
+      LOG(ERROR) << second_handler->DumpState();
+    }
+
+    for (const auto &it : whitelist) {
+      it.ApplyToInsn(isel_name, func1_state);
+      it.ApplyToInsn(isel_name, func2_state);
+    }
+
+    auto are_equal =
+        std::memcmp(func1_state, func2_state, sizeof(X86State)) == 0 &&
+        memory_state_eq;
+
+
+    return {init_state, this->DumpState(func1_state),
+            this->DumpState(func2_state), are_equal};
+  }
+};
+
+struct TestCase {
+  uint64_t addr;
+  std::string bytes;
+};
+
+namespace llvm::json {
+bool fromJSON(const Value &E, TestCase &Out, Path P) {
+  auto byte_string = E.getAsString();
+  if (!byte_string.hasValue()) {
+    P.report("Expected hex string of instruction bytes");
+    return false;
+  }
+
+  auto bytes = llvm::fromHex(byte_string.getValue());
+
+  Out.bytes = bytes;
+  // Should maybe do something else here?
+  Out.addr = 0xdeadbe00;
+  return true;
+}
+};  // namespace llvm::json
+
+
+std::string test_case_name(std::string_view prefix, uint64_t test_cast_ctr) {
+  std::stringstream ss;
+  ss << prefix << "comp_func" << test_cast_ctr;
+  return ss.str();
+}
+
+// Returns true when test case succeeds
+bool runTestCase(const TestCase &tc, DifferentialModuleBuilder &diffbuilder,
+                 const std::vector<WhiteListInstruction> &whitelist,
+                 uint64_t ctr) {
+  LOG(INFO) << "Starting testcase: " << llvm::toHex(tc.bytes);
+  auto diff_mod = diffbuilder.build(
+      test_case_name("f1", ctr), test_case_name("f2", ctr), tc.bytes, tc.addr);
+
+  if (!diff_mod.has_value()) {
+    LOG(ERROR) << "Failed to lift " << std::hex << tc.addr << ": "
+               << llvm::toHex(tc.bytes);
+
+    if (FLAGS_stop_on_fail) {
+      LOG(FATAL) << "Failed to lift an insn";
+    }
+    return false;
+  }
+
+  auto end = diff_mod->GetModule()->getDataLayout().isBigEndian()
+                 ? llvm::support::endianness::big
+                 : llvm::support::endianness::little;
+  ComparisonRunner comp_runner(end);
+
+  if (FLAGS_should_dump_functions) {
+    LOG(INFO) << remill::LLVMThingToString(diff_mod->GetF1());
+    LOG(INFO) << remill::LLVMThingToString(diff_mod->GetF2());
+  }
+
+  for (uint64_t i = 0; i < FLAGS_num_iterations; i++) {
+    auto tc_result = comp_runner.SingleCmpRun(
+        tc.bytes.size(), diff_mod->GetF1(), diff_mod->GetF2(), whitelist,
```

Comment:
`diff_mod->GetFns` feels like a better design - it can return a two element tuple for example, so inside the callee (possibly a few callees later) you can unpack it.